src/webgl/buffer.js

```diff
@@ -185,17 +185,56 @@ export default class Buffer extends Resource {
     return this;
   }
 
-  // WEBGL2 ONLY: Reads data from buffer into an ArrayBuffer or SharedArrayBuffer.
+  //
+  /**
+   * WEBGL2 ONLY: Reads data from buffer into an ArrayBufferView or SharedArrayBuffer.
+   *
+   * @param {ArrayBufferView} dstData= -ArrayBufferView to copy elements into,
+   *    new ArrayBufferView allocated with correct type if not provided.
+   * @param {GLuint} length= -Element count to be copied, optimal value calculated
+   *    when not provided.
+   * @param {GLuint} srcByteOffset=0 Byte offset into buffer data from where data to be copied.
+   * @param {GLuint} dstOffset=0 Element offset into dstData to where data to be copied.
+   *
+   * @returns {ArrayBufferView} - Copies elements from this buffer and
+   *    returns supplied/allocated ArrayBufferView.
+   */
   getData({
     dstData,
+    length,
     srcByteOffset = 0,
-    dstOffset = 0,
-    length = 0
+    dstOffset = 0
   }) {
-    // TODO optimize dstData according to offset and length
-    dstData = dstData || new ArrayBuffer(this.bytes);
+    assertWebGL2Context(this.gl);
+
+    // Use GL_COPY_READ_BUFFER to avoid disturbing other targets and locking type
```

Comment:
This comment got detached from its code.

Reply:
This is a lot of preliminary calculations, should we create a helper function to perform the work?